python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,582 @@
+from typing import List, Union
+
+import numpy as np  # type: ignore
+
+import tvm
+from tvm import relax as rx
+from tvm.relax.struct_info import TensorStructInfo
+from tvm.runtime.container import tuple_object
+from tvm.relax.op import add, subtract, multiply, divide, sqrt
+from tvm.relax import Var, Function
+
+
+class Optimizer:
+    """Relax training optimizer. This class could generate relax Functions for optimizing specified
+    parameters, and store the states used in the optimization process, such as momentum.
+
+    Parameters
+    ----------
+    params : Union[Var, List[Var]]
+        The parameter or the list of parameters to optimize.
+
+        If params is None, it indicates params will be added later using add_params.
+
+        Parameters should all be Vars of floating point Tensors, including float32, float64,
+        float16, bfloat16, etc. Currently, all parameters should have the same dtype, and that dtype
+        will be used as the dtype of the optimizer states.
+    """
+
+    _param_list: List[Var]
+    _state: tvm.runtime.container.ADT
+    _dtype: str
+
+    def __init__(self, params: Union[Var, List[Var]]) -> None:
+        if params is None:
+            params = []
+        elif not isinstance(params, list):
+            params = [params]
+        self._state = None
+        self._dtype = None
+        self._check_params_and_dtype(params)
+        self._param_list = params
+        assert len(self._param_list) == len(
+            set(self._param_list)
+        ), "There exist duplicate Vars in the parameter list."
+
+    def _check_params_and_dtype(self, params: List[Var]):
+        for x in params:
+            assert isinstance(x, Var), "Not every parameter is Var."
+            assert isinstance(x.struct_info, TensorStructInfo), "Not every parameter is Tensor Var"
+            data_type = tvm.DataType(x.struct_info.dtype)
+            assert data_type.type_code in (
+                tvm.DataTypeCode.BFLOAT,
+                tvm.DataTypeCode.FLOAT,
+            ), "Pamameters must be of float dtype"
+            if self._dtype is None:
+                self._dtype = x.struct_info.dtype
+            else:
+                assert (
+                    self._dtype == x.struct_info.dtype
+                ), "All parameters should have the same dtype"
+
+    def add_params(self, params: Union[Var, List[Var]]):
+        """Append one parameter or a list of new parameters to the optimizer.
+
+        Parameters
+        ----------
+        params : Union[Var, List[Var]]
+            The parameter or the list of parameters to append.
+
+            Parameters should all be Vars of floating point Tensors, including float32, float64,
+            float16, bfloat16, etc. Currently, all parameters should have the same dtype, and that
+            dtype will be used as the dtype of the optimizer states.
+
+        Note
+        ----
+        This method can only be called before `opt.state` or `opt.get_function()`.
```

Comment:
Could you elaborate a bit about the reason of supporting adding more parameters? I’m curious about its use cases.

Reply:
Removed after discussion